packages/@react-aria/dialog/src/useDialog.ts

```diff
@@ -62,6 +67,8 @@ export function useDialog(props: AriaDialogProps, ref: RefObject<FocusableElemen
   }, [ref]);
 
   useOverlayFocusContain();
+  // TODO: keep in mind this will stop propagation and prevent the Esc handler in useOverlay from firing I believe
+  let {keyboardProps} = useKeyboard({onKeyDown, onKeyUp});
```

Comment:
Due to the nature of `data-react-aria-top-layer` and its special handling in useOverlay/usePopover, ESC won't close it. To resolve this I opted to handle Esc via `useSubMenuTrigger` sending a keydown handler to the Dialog, open to opinions